_posts/2022-10-01-arrow-parquet-encoding-part-3.md

```diff
@@ -0,0 +1,169 @@
+}
+```
+
+Documents of this format could be stored in this arrow schema
+
+```text
+Field(name: “a”, nullable: true, datatype: List(
+  Field(name: “element”, nullable: true, datatype: Int32),
+)
+Field(name: “b”), nullable: false, datatype: List(
+  Field(name: “element”, nullable: true, datatype: Struct[
+    Field(name: “b2”, nullable: false, datatype: Int32),
+    Field(name: “c2”, nullable: true, datatype: List(
+      Field(name: “element”, nullable: false, datatype: Int32)
+    ))
+  ])
+))
+```
+
+Documents of this format could be stored in this parquet schema
+
+```text
+message schema {
+  optional group a (LIST) {
+    repeated group list {
+      optional int32 element;
+    }
+  }
+  required group b (LIST) {
+    repeated group list {
+      optional group element {
+        required int32 c1;
+        optional group c2 (LIST) {
+          repeated group list {
+            required int32 element;
+          }
+        }
+      }
+    }
+  }
+}
+```
+
+As explained previously, Arrow chooses to represent this in a hierarchical fashion. To achieve this it stores a list of monotonically increasing integers called offsets in the parent ListArray, and stores all the values that appear in the lists in a single child array. Each consecutive pair of elements in this offset array identifies a slice of the child array for that array index.
+
+```text
+a: ListArray
+  Offsets: [0, 1, 1, 3]
+  Validity: [true, false, true]
+  Children:
+    element: PrimitiveArray
+      Buffer[0]: [1, ARBITRARY, ARBITRARY]
+      Validity: [true, false, false]
+b: ListArray
+  Offsets: [0, 2, 3, 4]
+  Children:
+    element: StructArray
+      Validity: [true, true, true, false]
+      Children:
+        c1: PrimitiveArray
+          Buffer[0]: [1, 1, 2, ARBITRARY]
+        c2: ListArray
+          Offsets: [0, 0, 2, 2, 2]
+          Validity: [false, true, false, ARBITRARY]
+          Children:
+            element: PrimitiveArray
+              Buffer[0]: [3, 4]
+```
+
+In order to encode lists, Parquet stores an integer repetition level in addition to a definition level. A repetition level identifies where in the hierarchy of repeated fields the current value is to be inserted. A value of 0 would imply a new list in the top-most repeated field, a value of 1 a new element within the top-most repeated field, a value of 2 a new element within the second top-most repeated field, and so on.
+
+Each repeated field also has a corresponding definition level, however, in this case rather than indicating a null value, they indicate an empty array.
+
+```text
+a:
+  Data Page:
+    Repetition Levels: encode([0, 0, 0, 1])
+    Definition Levels: encode([3, 0, 2, 2])
+    Values: encode([1])
+b.c1:
+  Data Page:
+    Repetition Levels: encode([0, 1, 0, 0])
+    Definition Levels: encode([2, 2, 2, 1])
+    Values: encode([1, 1, 2])
+b.c2
+  Data Page:
+    Repetition Levels: encode([0, 1, 2, 0, 0])
+    Definition Levels: encode([2, 3, 3, 2, 1])
```

Comment:
```suggestion
    Definition Levels: encode([2, 4, 4, 2, 1])
```

A definition level of 0 would be `b: []` a level of 1 would be`b: [null]`, a level of 2 would be `b: [{b2: null}]`, a level of 3 would be `b: [{b2: []}]`. I forgot that b2 needs a definition level to encode an empty list